lib/HLSL/DxilNoOptLegalize.cpp

```diff
@@ -120,10 +124,27 @@ class DxilNoOptSimplifyInstructions : public ModulePass {
               I->eraseFromParent();
               Changed = true;
             }
+          } else if (PHINode* Phi = dyn_cast<PHINode>(I)) {
+            if (Phi->getNumIncomingValues() == 1
+             && Phi->getType() == HandleType) {
+                Changed = true;
+                LcssaHandlePhis.push_back(Phi);
+            }
           }
         }
       }
     }
+
+    // Replace all single value phi nodes (these are inserted by lcssa) of
```

Comment:
I would also expect these should be cleaned up by resource legalization (part of `hlsl-dxil-lower-handle-for-lib`).  Maybe we are missing some cleanup for this case.

Reply:
That pass does run, but leaves the phi in place. Let me see if we can move the cleanup there.

Reply:
From what I can tell it looks like that pass does not try to remove phis for Handles, but just for the global variables that get fed into createHandleForLib.

At the time the pass runs the IR looks like this
```
@"\01?buf@@3V?$StructuredBuffer@US@@@@A" = external global %"class.StructuredBuffer<S>", align 4
@llvm.used = appending global [1 x i8*] [i8* bitcast (%"class.StructuredBuffer<S>"* @"\01?buf@@3V?$StructuredBuffer@US@@@@A" to i8*)], section "llvm.metadata"

; Function Attrs: convergent
declare void @dx.noop() #0

; Function Attrs: nounwind
define void @main(i32* noalias, i32) #1 {
entry:
  %2 = load %"class.StructuredBuffer<S>", %"class.StructuredBuffer<S>"* @"\01?buf@@3V?$StructuredBuffer@US@@@@A", !dbg !3, !noalias !10 ; line:6 col:12
  %3 = call i32 @dx.op.loadInput.i32(i32 4, i32 0, i32 0, i8 0, i32 undef), !dbg !13 ; line:12 col:9
  call void @dx.noop(), !dbg !13 ; line:12 col:9
  %cmp = icmp ult i32 0, %3, !dbg !14 ; line:16 col:16
  %tobool = icmp ne i1 %cmp, false, !dbg !14 ; line:16 col:16
  br label %do.body, !dbg !15 ; line:14 col:5

do.body:                                          ; preds = %do.cond, %entry
  call void @dx.noop(), !dbg !16 ; line:15 col:13
  %4 = call %dx.types.Handle @"dx.op.createHandleForLib.class.StructuredBuffer<S>"(i32 160, %"class.StructuredBuffer<S>" %2), !dbg !3 ; line:6 col:12
  %5 = call %dx.types.Handle @dx.op.annotateHandle(i32 216, %dx.types.Handle %4, %dx.types.ResourceProperties { i32 524, i32 4 }), !dbg !3 ; line:6 col:12
  call void @dx.noop() #1, !dbg !3, !noalias !10 ; line:6 col:12
  call void @dx.noop() #1, !dbg !17, !noalias !10 ; line:6 col:5
  call void @dx.noop(), !dbg !16 ; line:15 col:13
  br label %do.cond, !dbg !18 ; line:16 col:5

do.cond:                                          ; preds = %do.body
  %tobool1 = icmp ne i1 %tobool, false, !dbg !18 ; line:16 col:5
  br i1 %tobool1, label %do.body, label %do.end, !dbg !18 ; line:16 col:5

do.end:                                           ; preds = %do.cond
  %.lcssa = phi %dx.types.Handle [ %5, %do.cond ]
  %RawBufferLoad = call %dx.types.ResRet.i32 @dx.op.rawBufferLoad.i32(i32 139, %dx.types.Handle %.lcssa, i32 0, i32 0, i8 1, i32 4), !dbg !19 ; line:18 col:14
  %6 = extractvalue %dx.types.ResRet.i32 %RawBufferLoad, 0, !dbg !19 ; line:18 col:14
  call void @dx.noop(), !dbg !20 ; line:18 col:5
  call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 %6), !dbg !20 ; line:18 col:5
  ret void, !dbg !20 ; line:18 col:5
}
```
The problem is the `%5` that gets fed to the phi, but it looks like the pass is trying to fix up the uses of `@"\01?buf@@3V?$StructuredBuffer@US@@@@A"`


The `DxilLowerCreateHandleForLib` pass calls `LegalizeResourceUseHelper::runOnModule` which eventually calls `ColectResourceGVUsers` on the global variable for the resource. Inside `CollectResourceGVUsers` it stops recursing when it hits a use of the `createHandleForLib` function

https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/HLSL/DxilCondenseResources.cpp#L1093C9-L1093C9